### PR TITLE
feat(#8685utq2d): Implement Ontology Mapping Gap Detection

### PR DESCRIPTION
**Pull request by connorjcantrell**
This PR introduces functionality to load and process ontologies from their respective NuGet, parse DTDL, parse their mappings, and detect gaps in the mappings.

Key Features:
- Load ontologies for both `mapped.ontologies.core.dtdl` and `willowinc.ontology.dtdlv3` NuGet packages
- Parse respective mappings from `Ontologies.Mappings` source.
- Identify gaps in the mappings for both ontologies.
- Output missing gaps to a `gaps.json` file.

Changes:
- Main execution logic added in `main.py`.
- Introduced `DTDLHandler` class in `handler.py` to handle ontology processing.
- New `mappings` and `ontology` sub-modules for respective features.

### CODE REVIEW

**Comment by connorjcantrell** [comment]
@jbkoh This PR currently identifies which concepts are not represented in the mapping files. However, it doesn't identify gaps between the two mapping files. If you would like to include this ability, we could migrate https://github.com/mapped/contrast into this repository.

File: scripts/mapping_diagnostics/main.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        "willow": {
+            "missing_namespaces": sorted(list(missing_willow_namespaces)),
+            "missing_relationships": sorted(list(missing_willow_relationships)),
+            "missing_interfaces": sorted(list(missing_willow_interfaces))
+        }
+    }
+
+    with open('gaps.json', 'w') as file:
+        json.dump(gaps, file, indent=2)
+
```
Comment (jbkoh) [inline scripts/mapping_diagnostics/main.py:42]:
```python
if __name__ == '__main__':
  main()
```
is the common convention for this

**Review by jbkoh — CHANGES_REQUESTED** [review]
The code looks good overall. Please update the github action to include the output of this as an artifact.

**Review by jbkoh — CHANGES_REQUESTED** [review]
The code looks good overall. Please update the github action to include the output of this as an artifact.

**Comment by jbkoh** [comment]
just a quick comment that `modules` is an unnecessary level.

File: .github/workflows/Ontology_Mappings_Diagnostics.yml [hunk 2]
```diff
@@ -17,8 +17,7 @@ jobs:
         path: ~/.cache/pypoetry/virtualenvs
         key: ${{ runner.os }}-poetry-${{ hashFiles('poetry.lock') }}
     - run: poetry --version
-    - run: poetry install --no-interaction --no-root
-      if: steps.cache-deps.outputs.cache-hit != 'true'
+    - run: poetry install
```
Comment (connorjcantrell) [inline .github/workflows/Ontology_Mappings_Diagnostics.yml:20]: @jbkoh I'm having trouble with using my installed dependencies within the GH action. Poetry shell doesn't work.

**Review by connorjcantrell — COMMENTED** [review]
fixed

**Review by jbkoh — APPROVED** [review]

**Pull request merged**